eventauth.go

```diff
@@ -86,6 +86,40 @@ func StateNeededForAuth(events []Event) (result StateNeeded) {
 	return
 }
 
+// TuplesFromStateNeeded returns the needed state key tuples for performing auth on an event.
+func TuplesFromStateNeeded(needed StateNeeded) (res []StateKeyTuple) {
+	if needed.Create {
+		res = append(res, StateKeyTuple{"m.room.create", ""})
+	}
+	if needed.JoinRules {
+		res = append(res, StateKeyTuple{"m.room.join_rules", ""})
+	}
+	if needed.PowerLevels {
+		res = append(res, StateKeyTuple{"m.room.power_levels", ""})
+	}
+	for _, userID := range needed.Member {
+		res = append(res, StateKeyTuple{"m.room.member", userID})
+	}
+	for _, token := range needed.ThirdPartyInvite {
+		res = append(res, StateKeyTuple{"m.room.third_party_invite", token})
+	}
+	return
+}
+
+// AuthEventReferencesFromStateNeeded returns a list of event references which can be used under the 'auth_events' key
+// of a matrix event. 'authEvents' should contain all the current state events which may be used for auth, as this
+// function may require their event reference.
+func AuthEventReferencesFromStateNeeded(authEvents map[StateKeyTuple]*Event, needed StateNeeded) (refs []EventReference) {
+	tuples := TuplesFromStateNeeded(needed)
+	for _, t := range tuples {
+		e, ok := authEvents[t]
+		if ok && e != nil {
```

Comment:
Technically the `ok` check is unnecessary here because `e` will be `nil` if the tuple isn't present in the map.